autopush/web/truss_handler.py

```diff
@@ -0,0 +1,225 @@
+        else:
+            d = self._create_endpoint()
+            d.addCallback(self._return_endpoint, new_uaid)
+            d.addErrback(self._response_err)
+
+    @threaded_validate(TrussSchema)
+    def put(self, *args):
+        """HTTP PUT
+
+        Update router type/data for a UAID.
+
+        """
+        self.start_time = time.time()
+
+        self.uaid = self.valid_input['uaid']
+        router = self.valid_input['router']
+        self.add_header("Content-Type", "application/json")
+        d = Deferred()
+        d.addCallback(router.register, router_data=self.valid_input['params'],
+                      app_id=self.valid_input['router_token'],
+                      uri=self.request.uri)
+        d.addCallback(self._save_router_data, self.valid_input['router_type'])
+        d.addCallback(self._success)
+        d.addErrback(self._router_fail_err)
+        d.addErrback(self._response_err)
+        d.callback(self.valid_input['uaid'])
+
+    def _delete_channel(self, uaid, chid):
+        message = self.ap_settings.message
+        if not message.unregister_channel(str(uaid), chid):
+            raise ItemNotFound("ChannelID not found")
+
+    def _delete_uaid(self, uaid, router):
+        self.log.info(format="Dropping User", code=101,
+                      uaid_hash=hasher(str(uaid)))
+        if not router.drop_user(str(uaid)):
+            raise ItemNotFound("UAID not found")
+
+    def _register_channel(self, router_data=None):
+        self.ap_settings.message.register_channel(self.uaid,
+                                                  self.chid)
+        endpoint = self.ap_settings.make_endpoint(self.uaid,
+                                                  self.chid,
+                                                  self.app_server_key)
+        return endpoint, router_data
+
+    @threaded_validate(TrussSchema)
+    def delete(self, *args):
+        """HTTP DELETE
+
+        Delete all pending records for the given channel or UAID
+
+        """
+        if self.valid_input['chid']:
+            # mark channel as dead
+            self.ap_settings.metrics.increment("updates.client.unregister",
+                                               tags=self.base_tags())
+            d = deferToThread(self._delete_channel,
+                              self.valid_input['uaid'],
+                              self.valid_input['chid'])
+            d.addCallback(self._success)
+            d.addErrback(self._chid_not_found_err)
+            d.addErrback(self._response_err)
+            return d
+        # nuke uaid
+        d = deferToThread(self._delete_uaid, self.valid_input['uaid'],
+                          self.ap_settings.router)
+        d.addCallback(self._success)
+        d.addErrback(self._uaid_not_found_err)
+        d.addErrback(self._response_err)
+        return d
+
+    def new_id(self):
+        """return a UUID4 value
+
+        This is mostly to deal with @patch not always working for testing
```

Comment:
We have plenty of tests that patch this successfully though? Which test specifically, let's fix that instead of add code like this?